File: app/src/main/java/com/vectras/vm/setupwizard/FirstRunPermissionOrchestrator.java
package com.vectras.vm.setupwizard;

import android.app.Activity;
import android.os.Build;

import androidx.annotation.NonNull;

import com.vectras.vm.R;
import com.vectras.vm.utils.PermissionUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class FirstRunPermissionOrchestrator {
    public enum PermissionStatus {
        PENDING,
        GRANTED,
        SKIPPED,
        FAILED
    }

    public static class PermissionUiModel {
        public final String capability;
        public final String title;
        public final String description;
        public final boolean essential;
        public final boolean canOpenSettings;
        public final boolean canRetry;
        public PermissionStatus status;

        PermissionUiModel(String capability,
                          String title,
                          String description,
                          boolean essential,
                          boolean canOpenSettings,
                          boolean canRetry,
                          PermissionStatus status) {
            this.capability = capability;
            this.title = title;
            this.description = description;
            this.essential = essential;
            this.canOpenSettings = canOpenSettings;
            this.canRetry = canRetry;
            this.status = status;
        }
    }

    public static final String CAPABILITY_STORAGE = "storage";

    private final Activity activity;
    private final List<PermissionUiModel> models;

    public FirstRunPermissionOrchestrator(@NonNull Activity activity) {
        this.activity = activity;
        this.models = new ArrayList<>();
        models.add(new PermissionUiModel(
                CAPABILITY_STORAGE,
                activity.getString(R.string.allow_access_to_storage),
                activity.getString(R.string.you_need_to_allow_access_to_the_storage_to_continue),
                true,
                true,
                true,
                PermissionStatus.PENDING
        ));
        refresh();
    }

    public void refresh() {
        for (PermissionUiModel model : models) {
            if (CAPABILITY_STORAGE.equals(model.capability)) {
                boolean granted = isStorageGranted();
                if (granted) {
                    model.status = PermissionStatus.GRANTED;
                } else if (canSkipStorageByBusinessRule()) {
                    model.status = PermissionStatus.SKIPPED;
                } else if (model.status != PermissionStatus.FAILED) {
                    model.status = PermissionStatus.PENDING;
                }
            }
        }
    }

    public void markFailed(@NonNull String capability) {
        for (PermissionUiModel model : models) {
            if (capability.equals(model.capability)) {
                model.status = PermissionStatus.FAILED;
                return;
            }
        }
    }

    public boolean isEssentialResolved() {
        refresh();
        for (PermissionUiModel model : models) {
            if (!model.essential) {
                continue;
            }
            if (model.status != PermissionStatus.GRANTED && model.status != PermissionStatus.SKIPPED) {
                return false;
            }
        }
        return true;
    }

    public List<PermissionUiModel> getUiModel() {
        refresh();
        return Collections.unmodifiableList(models);
    }

    private boolean isStorageGranted() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            return PermissionUtils.isAllFilesAccessGranted() || PermissionUtils.hasPersistedTreePermission(activity);
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            return PermissionUtils.hasPersistedTreePermission(activity);
        }
        return PermissionUtils.storagepermission(activity, false);
    }

    private boolean canSkipStorageByBusinessRule() {
        return Build.VERSION.SDK_INT == Build.VERSION_CODES.Q;
    }
}

File: app/src/main/java/com/vectras/vm/setupwizard/FirstRunPermissionOrchestrator.kt
package com.vectras.vm.setupwizard

import android.app.Activity
import android.os.Build
import com.vectras.qemu.MainSettingsManager
import com.vectras.vm.utils.PermissionUtils

class FirstRunPermissionOrchestrator(private val activity: Activity) {

    enum class Capability {
        STORAGE_SAF,
        NOTIFICATIONS,
        BATTERY_OPTIMIZATION,
        OVERLAY,
        MEDIA_ACCESS
    }

    enum class StepState {
        PENDING,
        GRANTED,
        SKIPPED,
        FAILED
    }

    data class CapabilityStep(
        val capability: Capability,
        val required: Boolean,
        val state: StepState
    )

    private val orderedCapabilities = listOf(
        Capability.STORAGE_SAF,
        Capability.NOTIFICATIONS,
        Capability.BATTERY_OPTIMIZATION,
        Capability.OVERLAY,
        Capability.MEDIA_ACCESS
    )

    fun getSteps(): List<CapabilityStep> {
        return orderedCapabilities
            .filter { isApplicable(it) }
            .map {
                CapabilityStep(
                    capability = it,
                    required = isRequired(it),
                    state = resolveState(it)
                )
            }
    }

    fun getNextPendingRequired(): Capability? {
        return getSteps().firstOrNull { it.required && it.state == StepState.PENDING }?.capability
    }

    fun areRequiredCapabilitiesGranted(): Boolean {
        return getSteps().none { it.required && it.state != StepState.GRANTED }
    }

    fun refreshPersistedStates() {
        for (capability in orderedCapabilities) {
            if (!isApplicable(capability)) {
                persistState(capability, StepState.SKIPPED)
                continue
            }

            if (isCapabilityGranted(capability)) {
                persistState(capability, StepState.GRANTED)
            } else {
                val current = getPersistedState(capability)
                if (current != StepState.FAILED && current != StepState.SKIPPED) {
                    persistState(capability, StepState.PENDING)
                }
            }
        }
    }

    fun markSkipped(capability: Capability) {
        persistState(capability, StepState.SKIPPED)
    }

    fun markFailed(capability: Capability) {
        persistState(capability, StepState.FAILED)
    }

    fun markGranted(capability: Capability) {
        persistState(capability, StepState.GRANTED)
    }

    private fun resolveState(capability: Capability): StepState {
        if (isCapabilityGranted(capability)) {
            persistState(capability, StepState.GRANTED)
            return StepState.GRANTED
        }

        return getPersistedState(capability)
    }

    private fun isCapabilityGranted(capability: Capability): Boolean {
        return when (capability) {
            Capability.STORAGE_SAF -> PermissionUtils.hasStorageCapability(activity)
            Capability.NOTIFICATIONS -> PermissionUtils.hasNotificationCapability(activity)
            Capability.BATTERY_OPTIMIZATION -> PermissionUtils.isBatteryOptimizationIgnored(activity)
            Capability.OVERLAY -> PermissionUtils.hasOverlayCapability(activity)
            Capability.MEDIA_ACCESS -> PermissionUtils.hasMediaReadCapability(activity)
        }
    }

    private fun isRequired(capability: Capability): Boolean {
        return capability == Capability.STORAGE_SAF
    }

    private fun isApplicable(capability: Capability): Boolean {
        return when (capability) {
            Capability.NOTIFICATIONS -> Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU
            Capability.MEDIA_ACCESS -> Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU
            Capability.OVERLAY -> {
                val vmUi = MainSettingsManager.getVmUi(activity)
                vmUi.equals("X11", ignoreCase = true) || vmUi.equals("VNC", ignoreCase = true)
            }
            else -> true
        }
    }

    private fun keyFor(capability: Capability): String {
        return "firstRunPermissionState_${capability.name}"
    }

    private fun persistState(capability: Capability, state: StepState) {
        MainSettingsManager.setFirstRunPermissionState(activity, keyFor(capability), state.name)
    }

    private fun getPersistedState(capability: Capability): StepState {
        val persisted = MainSettingsManager.getFirstRunPermissionState(activity, keyFor(capability))
        return try {
            StepState.valueOf(persisted)
        } catch (_: IllegalArgumentException) {
            StepState.PENDING
        }
    }
